packages/server-wallet/src/chain-service/chain-service.ts

```diff
@@ -325,19 +328,28 @@ export class ChainService implements ChainServiceInterface {
     this.channelToSubscribers.set(channelId, []);
   }
 
+  /**
+   * WARNING: potential for a race condition due to:
+   * 1. Read class state (this.finalizingChannels)
+   * 2. Await a promise (getChannelStorage)
+   * 3. Assume that the state in step 1 has not changed.
+   *
+   * If many onBlockMined are executed in parallel, step 3 assumption does not hold
+   * PQueue is used to avoid these race conditions.
+   */
   private async onBlockMined(block: providers.Block): Promise<void> {
     if (!this.finalizingChannels.length) return;
 
     const finalizingChannel = this.finalizingChannels[0];
```

Comment:
Nit: it is a bit confusing to me that this function recurses into itself. When we enter a second time, it may not be that a block has just been mined -- just that when we were found a hit in the `finalizingChannels` array we decided to go again for some reason. 

It seems better to me to have a function named `checkFinalizingChannels` or similar that is called *once* inside `onBlockMined`, and have *that* function recurse instead.

Reply:
d6649c8ad